File: gtci/11-tree-breadth-first-search/cpp/1-sum-of-binary-tree-bfs.cpp
#include <cassert>
#include <iostream>
#include <queue>

#include "tree_utils.h"

class Solution {
    public:
    int sumOfBinaryTree(TreeNode* root) {
        if (root == nullptr)
            return 0;

        int sum = 0;

        std::queue<TreeNode*> queue;
        queue.push(root);

        while (!queue.empty()) {
            TreeNode* node = queue.front();
            queue.pop();

            sum += node->val;

            if (node->left != nullptr)
                queue.push(node->left);
            
            if (node->right != nullptr)
                queue.push(node->right);
        }

        return sum;
    }
};

int main() {
    Solution sol = Solution();

    TreeNode* root1 = new TreeNode(1);
    root1->left = new TreeNode(2);
    root1->right = new TreeNode(3);
    assert(sol.sumOfBinaryTree(root1) == 6);
    
    TreeNode* root2 = new TreeNode(4);
    root2->left = new TreeNode(9);
    root2->right = new TreeNode(7);
    root2->left->left = new TreeNode(2);
    root2->left->right = new TreeNode(6);
    assert(sol.sumOfBinaryTree(root2) == 28);

    TreeNode* root3 = new TreeNode(10);
    root3->left = new TreeNode(5);
    root3->right = new TreeNode(3);
    root3->left->left = new TreeNode(7);
    root3->right->right = new TreeNode(9);
    assert(sol.sumOfBinaryTree(root3) == 34);

    std::cout << "All test cases passed." << std::endl;

    delete root1;
    delete root2;
    delete root3;

    return 0;
}

File: gtci/11-tree-breadth-first-search/python3/1-sum-of-binary-tree-bfs.py
from collections import deque


class TreeNode:
    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def sum_of_binary_tree(self, root):
        if root is None:
            return 0

        sum = 0

        queue = deque()
        queue.append(root)

        while queue:
            node = queue.popleft()

            sum += node.val

            if node.left is not None:
                queue.append(node.left)

            if node.right is not None:
                queue.append(node.right)

        return sum


if __name__ == "__main__":
    sol = Solution()

    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    assert sol.sum_of_binary_tree(root) == 6

    root = TreeNode(4)
    root.left = TreeNode(9)
    root.right = TreeNode(7)
    root.left.left = TreeNode(2)
    root.left.right = TreeNode(6)
    assert sol.sum_of_binary_tree(root) == 28

    root = TreeNode(10)
    root.left = TreeNode(5)
    root.right = TreeNode(3)
    root.left.left = TreeNode(7)
    root.right.right = TreeNode(9)
    assert sol.sum_of_binary_tree(root) == 34

    print("All test cases passed.")
